File: new/has_duplicates_sorting.cpp
/*
nums.begin()  → &nums[0]
nums.end()    → &nums[nums.size()]   // one past the end
*/


class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        sort(nums.begin(), nums.end()); // O(nlogn)

        for (int i = 1; i < nums.size(); i++) { // O(n)
            if (nums[i] == nums[i - 1]) {
                return true;
            }
        }

        return false;
        // time : O(nlogn) + O(n) = O(nlogn)
        // space : O(1) because of in place sorting, if not O(n)
    }
};

File: new/python/has_duplicates_sorting.py
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        # using sorting
        # step 1: sort
        nums.sort()  # O(nlogn)

        # step 2: compare
        for i in range(1, len(nums)):  # O(n)
            if nums[i] == nums[i - 1]:
                return True

        return False
        # time complexity : O(nlogn) + O(n) = O(nlogn)
        # space complexity : if inplace sorting = O(1) else O(n)
